### ISSUE TITLE
Support React Native

### ISSUE THREAD

**Issue by gutenye** [jaydenseric/apollo-upload-client#10]
React Native is [almost same](https://github.com/facebook/react-native/blob/master/Libraries/Network/FormData.js#L28), using `fetch`, except no `windows` object.

**Comment by jaydenseric** [comment]
I don't have experience with react-native, but it it seems like a good idea to support it. Are you able to share what specifically doesn't work, or needs to be changed?

**Comment by joncursi** [comment]
I tried adding this to a RN project and realized as well that this package is coupled to web apps at the moment. Can't get RN to compile.

![screen shot 2017-05-24 at 10 04 10 pm](https://cloud.githubusercontent.com/assets/4974609/26432900/131f2ab6-40cd-11e7-914f-0f199441a210.png)

**Comment by jaydenseric** [comment]
Thanks all for the various pull requests and suggestions, they have helped me to understand most of what needs to change.

There is a new [`support-react-native` branch](https://github.com/jaydenseric/apollo-upload-client/tree/support-react-native).

I have tested it for web and everything seems to work the same, just need to make sure it actually works on React Native. If anyone could test it out and report issues, that would be a big help.

If you would like to submit a pull request to it, make sure to select that branch and not master.

One thing I am not super happy about is the possibility of false positives when identifying React Native file objects [here](https://github.com/jaydenseric/apollo-upload-client/blob/support-react-native/src/helpers.js#L23); the check just looks for the presence of relevant properties. It is conceivable that people might have nodes other than files with those properties.

**Comment by giautm** [comment]
Please notify user need to be install there modules before use.
```
npm - --save-dev babel-plugin-transform-runtime babel-preset-env babel-preset-stage-0
```
or
```
yarn add --dev babel-plugin-transform-runtime babel-preset-env babel-preset-stage-0
```

**Comment by giautm** [comment]
On trying install module via git. Then
```
yarn add https://github.com/jaydenseric/apollo-upload-client.git#support-react-native
yarn add --dev babel-plugin-transform-runtime babel-preset-env babel-preset-stage-0
```
AND
```
cd ./node_modules/apollo-upload-client
yarn build
```
Reload the app and getting:
![img_0592 2](https://user-images.githubusercontent.com/12751435/27426069-a6378fca-5764-11e7-9cf3-5f2a9cb4b897.PNG)

Install via npm or git (with dist folder committed) will work fine. I don't know why. :(

**Comment by jaydenseric** [comment]
That workaround to install the dev deps in your app will work, but that is only necessary (from my understanding) due to recent bugs in the React Native ecosystem.

About the git install thing, I just learned this about npm myself recently: Installs using Git only has the committed files in the repo to pull down, due to gitignore. npm installs for the last several major versions no longer run prepublish scripts. There is huge discussion in npm github issues about that frustration. Because npm link is horribly buggy, testing forks or updates is really inconvenient.

**Comment by jaydenseric** [comment]
To test the branch, try:

1. Cloning or downloading it.
2. Run `npm link` in it's directory.
3. In your project, run `npm link apollo-upload-client`.

To undo the above later, run the same commands in reverse order, with `unlink` in place of `link`.

**Comment by jaydenseric** [comment]
@gutenye on the `$export is not a function` issue, perhaps it relates to this: https://github.com/facebook/react-native/issues/4062#issuecomment-258219231.

**Comment by giautm** [comment]
Didn't work, https://github.com/jaydenseric/apollo-upload-client/issues/10#issuecomment-310324292 Huhu. :(

Can you make a npm release with tag #support-react-native ?

**Comment by jaydenseric** [comment]
Removed the feature branch, we will work on this on `master` for the next version.

Try installing `apollo-upload-client@4.1.0-alpha.1`.

**Comment by joncursi** [comment]
Just installed the latest alpha. Hit an issue on the `babel-plugin-transform-runtime` dependency.

![screen shot 2017-06-22 at 5 54 19 am](https://user-images.githubusercontent.com/4974609/27428194-530f82bc-570f-11e7-8ff8-5f1e4bfd3242.png)

Was able to get past this by installing that dependency as a dep on my own project. However, now I'm hitting:

![screen shot 2017-06-22 at 5 57 31 am](https://user-images.githubusercontent.com/4974609/27428353-c6d6090a-570f-11e7-9036-044c1688cb25.png)

**Comment by joncursi** [comment]
FYI - solved the above by also installing `babel-preset-env`. Project compiles now. Testing functionality...

**Comment by giautm** [comment]
@jaydenseric : I got an error with 4.1.0-alpha-1
```
 Can't find variable: File
```
![image](https://user-images.githubusercontent.com/12751435/27428609-cd94226a-576c-11e7-8ec2-934611830a79.png)

**Comment by giautm** [comment]
Add `typeof File !== 'undefined'` at here https://github.com/jaydenseric/apollo-upload-client/blob/master/src/helpers.js#L21
And `typeof FileList !== 'undefined'` at https://github.com/jaydenseric/apollo-upload-client/blob/master/src/helpers.js#L38

**Comment by jaydenseric** [comment]
@gutenye good stuff. Any other obvious issues before I push a new version?

**Comment by giautm** [comment]
@jaydenseric : After i modify bundled source and reload, very things work okay.

Ps: You always tag wrong persion. :D

**Comment by jaydenseric** [comment]
Sorry!

Published `apollo-upload-client@4.1.0-alpha.2`.

**Comment by giautm** [comment]
@jaydenseric I can confirm it work on react-native. But, remember to remind user to do this https://github.com/jaydenseric/apollo-upload-client/issues/10#issuecomment-310321318

**Comment by joncursi** [comment]
I was able to get the code to execute on the React Native side, but ran into issues on the server side (I'm using Meteor on the backend to host an Express GQL server):

```
W20170622-06:24:22.274(-4)? (STDERR) Error
W20170622-06:24:22.274(-4)? (STDERR)     at readStream (/Users/jcursi/Sites/joncursi/redbird-web/node_modules/raw-body/index.js:196:17)
W20170622-06:24:22.274(-4)? (STDERR)     at getRawBody (/Users/jcursi/Sites/joncursi/redbird-web/node_modules/raw-body/index.js:106:12)
W20170622-06:24:22.274(-4)? (STDERR)     at read (/Users/jcursi/Sites/joncursi/redbird-web/node_modules/body-parser/lib/read.js:76:3)
W20170622-06:24:22.275(-4)? (STDERR)     at jsonParser (/Users/jcursi/Sites/joncursi/redbird-web/node_modules/body-parser/lib/types/json.js:127:5)
W20170622-06:24:22.275(-4)? (STDERR)     at Layer.handle [as handle_request] (/Users/jcursi/Sites/joncursi/redbird-web/node_modules/express/lib/router/layer.js:95:5)
W20170622-06:24:22.275(-4)? (STDERR)     at trim_prefix (/Users/jcursi/Sites/joncursi/redbird-web/node_modules/express/lib/router/index.js:317:13)
W20170622-06:24:22.276(-4)? (STDERR)     at /Users/jcursi/Sites/joncursi/redbird-web/node_modules/express/lib/router/index.js:284:7
W20170622-06:24:22.279(-4)? (STDERR)     at Function.process_params (/Users/jcursi/Sites/joncursi/redbird-web/node_modules/express/lib/router/index.js:335:12)
W20170622-06:24:22.279(-4)? (STDERR)     at next (/Users/jcursi/Sites/joncursi/redbird-web/node_modules/express/lib/router/index.js:275:10)
W20170622-06:24:22.280(-4)? (STDERR)     at /Users/jcursi/Sites/joncursi/redbird-web/node_modules/apollo-upload-server/dist/apollo-upload-server.js:74:52
```

**Comment by jaydenseric** [comment]
@giautm in a React Native project, would the additional devDependencies need to be installed with `--save`, or `--save-dev`? If you used `--save-dev` would the RN packager error for production installs of your project? Or do you only build locally, and deploy static builds? I don't know the RN ecosystem.

Once we work it out I will update the readme.

**Comment by jaydenseric** [comment]
@joncursi assuming the client is sending the right payload, that might be something to explore in `apollo-upload-server`. It looks like it is upset with the `return next()` [here](https://github.com/jaydenseric/apollo-upload-server/blob/v2.0.2/src/index.js#L55)?

**Comment by joncursi** [comment]
@jaydenseric I just noticed that. Maybe I'm doing it wrong on the client. @giautm how are you getting at / uploading your image? Are you using [react-native-fs](https://github.com/itinance/react-native-fs) to pick it up from the filesystem? Mind sharing a small code sample?

**Comment by giautm** [comment]
@jaydenseric : Every things will bundle at build step to `android.bundle.js.` or `android.ios.js`. Just install them to `devDependencies`.

@joncursi : I just pass a variable to input like: 
```
        return mutate({
          variables: {
            input: {
              ...input,
              clientMutationId,
              signature: input.signature
              photos: input.photos,
            },
          },

```
`input.photos` is arrray of shape: `{ uri: String, type: String, name: String}`

**Comment by giautm** [comment]
@joncursi : Variable look like this:
```
{
    "input": {
        "orderId": "VHJhbnNwb3J0T3JkZXI6NTkyYmUyNDRjZTRmYTkwYmJlYzg0Y2E3",
        "appointment": null,
        "cause": 8,
        "comment": null,
        "signature": [
            {
                "name": "file_0",
                "type": "image/png",
                "uri": "/Users/admin/Library/Developer/CoreSimulator/Devices/C02A708F-7003-4747-BB09-AA35530AFAC8/data/Containers/Data/Application/44243AD1-906B-4C2E-AC28-6D700BA99609/tmp/ReactABI17_0_0Native/CA1AD1AC-2B6D-47EB-9AE3-528A537E8AC7.png"
            }
        ],
        "photos": [],
        "timestamp": "2017-06-20T15:48:34.832Z",
        "clientMutationId": 1497973714832
    }
}
```

**Comment by giautm** [comment]
> One thing I am not super happy about is the possibility of false positives when identifying React Native file objects here; the check just looks for the presence of relevant properties. It is conceivable that people might have nodes other than files with those properties.

I think we can implement a `FileInput` for react-native like this:
```
// file-input.js

export class FileInput {
  constructor({ name, type, uri } = {}) {
    this.name = name;
    this.type = type;
    this.uri = uri;
  }

  static fromArray = (array, defaultType = 'image/png') => Array.isArray(array)
    ? array.map(({ name, type, uri }, index) => new FileInput({
        name: name || `file_${index}`,
        type: type || defaultType,
        uri,
      }))
    : null;
}

export isFile = node => typeof File !== 'undefined'
  ? node instanceof File
  : node instanceof FileInput
```

**Comment by jaydenseric** [comment]
Support has landed in v4.1.0.

Bugs or suggestions can be raised on new issues.

**Comment by zackify** [comment]
Added a pr to fix having to install dev dependencies in RN, a high priority issue imo #23
